// cncjs/dist/cncjs/server/controllers/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WRITE_SOURCE_SERVER = exports.WRITE_SOURCE_SENDER = exports.WRITE_SOURCE_FEEDER = exports.WRITE_SOURCE_CLIENT = exports.GLOBAL_OBJECTS = void 0;
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects
var GLOBAL_OBJECTS = {
  // Function properties
  parseFloat: parseFloat,
  parseInt: parseInt,
  // Fundamental objects
  Object: Object,
  Function: Function,
  Boolean: Boolean,
  // Numbers and dates
  Number: Number,
  Math: Math,
  Date: Date,
  // Text processing
  String: String,
  RegExp: RegExp,
  // Structured data
  JSON: JSON
}; // Write Source

exports.GLOBAL_OBJECTS = GLOBAL_OBJECTS;
var WRITE_SOURCE_CLIENT = 'client';
exports.WRITE_SOURCE_CLIENT = WRITE_SOURCE_CLIENT;
var WRITE_SOURCE_SERVER = 'server';
exports.WRITE_SOURCE_SERVER = WRITE_SOURCE_SERVER;
var WRITE_SOURCE_FEEDER = 'feeder';
exports.WRITE_SOURCE_FEEDER = WRITE_SOURCE_FEEDER;
var WRITE_SOURCE_SENDER = 'sender';
exports.WRITE_SOURCE_SENDER = WRITE_SOURCE_SENDER;
