// cncjs/dist/cncjs/server/controllers/Grbl/GrblLineParserResultSettings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _trim2 = _interopRequireDefault(require("lodash/trim"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var GrblLineParserResultSettings = /*#__PURE__*/function () {
  function GrblLineParserResultSettings() {
    _classCallCheck(this, GrblLineParserResultSettings);
  }

  _createClass(GrblLineParserResultSettings, null, [{
    key: "parse",
    value: function parse(line) {
      var r = line.match(/^(\$[^=]+)=([^ ]*)\s*(.*)/);

      if (!r) {
        return null;
      }

      var payload = {
        name: r[1],
        value: r[2],
        message: (0, _trim2["default"])(r[3], '()')
      };
      return {
        type: GrblLineParserResultSettings,
        payload: payload
      };
    }
  }]);

  return GrblLineParserResultSettings;
}();

var _default = GrblLineParserResultSettings;
exports["default"] = _default;
