// webappengine/dist/app/config/settings.production.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _crypto = require('crypto');

var _crypto2 = _interopRequireDefault(_crypto);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _urljoin = require('../lib/urljoin');

var _urljoin2 = _interopRequireDefault(_urljoin);

var _package = require('../../../package.json');

var _package2 = _interopRequireDefault(_package);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// hashedVersion
var hashedVersion = function (version) {
    var algorithm = 'sha1';
    var buf = String(version);
    var hash = _crypto2.default.createHash(algorithm).update(buf).digest('hex');
    return hash.substr(0, 8); // 8 digits
}(_package2.default.version);

var maxAge = 365 * 24 * 60 * 60 * 1000; // one year

exports.default = {
    hashedVersion: hashedVersion,
    assets: {
        // web
        web: {
            routes: [// with trailing slash
            (0, _urljoin2.default)(hashedVersion, '/'), // hashed route
            '/' // fallback
            ],
            path: _path2.default.resolve(__dirname, '..', '..', 'web'),
            maxAge: maxAge
        }
    },
    cluster: {
        // note. node-inspector cannot debug child (forked) process
        enable: false,
        maxWorkers: _os2.default.cpus().length || 1
    },
    winston: {
        // https://github.com/winstonjs/winston#logging-levels
        level: 'info'
    }
};
