// cncjs/dist/cncjs/server/api/api.events.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.update = exports.read = exports.fetch = exports.create = exports.__delete = void 0;

var _find = _interopRequireDefault(require("lodash/find"));

var _castArray = _interopRequireDefault(require("lodash/castArray"));

var _isPlainObject = _interopRequireDefault(require("lodash/isPlainObject"));

var _uuid = _interopRequireDefault(require("uuid"));

var _settings = _interopRequireDefault(require("../config/settings"));

var _logger = _interopRequireDefault(require("../lib/logger"));

var _configstore = _interopRequireDefault(require("../services/configstore"));

var _paging = require("./paging");

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var log = (0, _logger["default"])('api:events');
var CONFIG_KEY = 'events';

var getSanitizedRecords = function getSanitizedRecords() {
  var records = (0, _castArray["default"])(_configstore["default"].get(CONFIG_KEY, []));
  var shouldUpdate = false;

  for (var i = 0; i < records.length; ++i) {
    if (!(0, _isPlainObject["default"])(records[i])) {
      records[i] = {};
    }

    var record = records[i];

    if (!record.id) {
      record.id = _uuid["default"].v4();
      shouldUpdate = true;
    } // Defaults to true


    if (record.enabled === undefined) {
      record.enabled = true;
    } // Alias command


    if (!record.commands) {
      record.commands = record.command || '';
      delete record.command;
    }
  }

  if (shouldUpdate) {
    log.debug("update sanitized records: ".concat(JSON.stringify(records))); // Pass `{ silent changes }` will suppress the change event

    _configstore["default"].set(CONFIG_KEY, records, {
      silent: true
    });
  }

  return records;
};

var fetch = function fetch(req, res) {
  var records = getSanitizedRecords();
  var paging = !!req.query.paging;

  if (paging) {
    var _req$query = req.query,
        _req$query$page = _req$query.page,
        page = _req$query$page === void 0 ? 1 : _req$query$page,
        _req$query$pageLength = _req$query.pageLength,
        pageLength = _req$query$pageLength === void 0 ? 10 : _req$query$pageLength;
    var totalRecords = records.length;

    var _getPagingRange = (0, _paging.getPagingRange)({
      page: page,
      pageLength: pageLength,
      totalRecords: totalRecords
    }),
        _getPagingRange2 = _slicedToArray(_getPagingRange, 2),
        begin = _getPagingRange2[0],
        end = _getPagingRange2[1];

    var pagedRecords = records.slice(begin, end);
    res.send({
      pagination: {
        page: Number(page),
        pageLength: Number(pageLength),
        totalRecords: Number(totalRecords)
      },
      records: pagedRecords.map(function (record) {
        var _record = _objectSpread({}, record),
            id = _record.id,
            mtime = _record.mtime,
            enabled = _record.enabled,
            event = _record.event,
            trigger = _record.trigger,
            commands = _record.commands;

        return {
          id: id,
          mtime: mtime,
          enabled: enabled,
          event: event,
          trigger: trigger,
          commands: commands
        };
      })
    });
  } else {
    res.send({
      records: records.map(function (record) {
        var _record2 = _objectSpread({}, record),
            id = _record2.id,
            mtime = _record2.mtime,
            enabled = _record2.enabled,
            event = _record2.event,
            trigger = _record2.trigger,
            commands = _record2.commands;

        return {
          id: id,
          mtime: mtime,
          enabled: enabled,
          event: event,
          trigger: trigger,
          commands: commands
        };
      })
    });
  }
};

exports.fetch = fetch;

var create = function create(req, res) {
  var _req$body = _objectSpread({}, req.body),
      _req$body$enabled = _req$body.enabled,
      enabled = _req$body$enabled === void 0 ? true : _req$body$enabled,
      _req$body$event = _req$body.event,
      event = _req$body$event === void 0 ? '' : _req$body$event,
      _req$body$trigger = _req$body.trigger,
      trigger = _req$body$trigger === void 0 ? '' : _req$body$trigger,
      _req$body$commands = _req$body.commands,
      commands = _req$body$commands === void 0 ? '' : _req$body$commands;

  if (!event) {
    res.status(_constants.ERR_BAD_REQUEST).send({
      msg: 'The "event" parameter must not be empty'
    });
    return;
  }

  if (!trigger) {
    res.status(_constants.ERR_BAD_REQUEST).send({
      msg: 'The "trigger" parameter must not be empty'
    });
    return;
  }

  if (!commands) {
    res.status(_constants.ERR_BAD_REQUEST).send({
      msg: 'The "commands" parameter must not be empty'
    });
    return;
  }

  try {
    var records = getSanitizedRecords();
    var record = {
      id: _uuid["default"].v4(),
      mtime: new Date().getTime(),
      enabled: !!enabled,
      event: event,
      trigger: trigger,
      commands: commands
    };
    records.push(record);

    _configstore["default"].set(CONFIG_KEY, records);

    res.send({
      id: record.id,
      mtime: record.mtime
    });
  } catch (err) {
    res.status(_constants.ERR_INTERNAL_SERVER_ERROR).send({
      msg: 'Failed to save ' + JSON.stringify(_settings["default"].rcfile)
    });
  }
};

exports.create = create;

var read = function read(req, res) {
  var id = req.params.id;
  var records = getSanitizedRecords();
  var record = (0, _find["default"])(records, {
    id: id
  });

  if (!record) {
    res.status(_constants.ERR_NOT_FOUND).send({
      msg: 'Not found'
    });
    return;
  }

  var _record3 = _objectSpread({}, record),
      mtime = _record3.mtime,
      enabled = _record3.enabled,
      event = _record3.event,
      trigger = _record3.trigger,
      commands = _record3.commands;

  res.send({
    id: id,
    mtime: mtime,
    enabled: enabled,
    event: event,
    trigger: trigger,
    commands: commands
  });
};

exports.read = read;

var update = function update(req, res) {
  var id = req.params.id;
  var records = getSanitizedRecords();
  var record = (0, _find["default"])(records, {
    id: id
  });

  if (!record) {
    res.status(_constants.ERR_NOT_FOUND).send({
      msg: 'Not found'
    });
    return;
  }

  var _req$body2 = _objectSpread({}, req.body),
      _req$body2$enabled = _req$body2.enabled,
      enabled = _req$body2$enabled === void 0 ? record.enabled : _req$body2$enabled,
      _req$body2$event = _req$body2.event,
      event = _req$body2$event === void 0 ? record.event : _req$body2$event,
      _req$body2$trigger = _req$body2.trigger,
      trigger = _req$body2$trigger === void 0 ? record.trigger : _req$body2$trigger,
      _req$body2$commands = _req$body2.commands,
      commands = _req$body2$commands === void 0 ? record.commands : _req$body2$commands; // Skip validation for "enabled", "event", "trigger", and "commands"


  try {
    record.mtime = new Date().getTime();
    record.enabled = Boolean(enabled);
    record.event = String(event || '');
    record.trigger = String(trigger || '');
    record.commands = String(commands || ''); // Remove deprecated parameter

    if (record.command !== undefined) {
      delete record.command;
    }

    _configstore["default"].set(CONFIG_KEY, records);

    res.send({
      id: record.id,
      mtime: record.mtime
    });
  } catch (err) {
    res.status(_constants.ERR_INTERNAL_SERVER_ERROR).send({
      msg: 'Failed to save ' + JSON.stringify(_settings["default"].rcfile)
    });
  }
};

exports.update = update;

var __delete = function __delete(req, res) {
  var id = req.params.id;
  var records = getSanitizedRecords();
  var record = (0, _find["default"])(records, {
    id: id
  });

  if (!record) {
    res.status(_constants.ERR_NOT_FOUND).send({
      msg: 'Not found'
    });
    return;
  }

  try {
    var filteredRecords = records.filter(function (record) {
      return record.id !== id;
    });

    _configstore["default"].set(CONFIG_KEY, filteredRecords);

    res.send({
      id: record.id
    });
  } catch (err) {
    res.status(_constants.ERR_INTERNAL_SERVER_ERROR).send({
      msg: 'Failed to save ' + JSON.stringify(_settings["default"].rcfile)
    });
  }
};

exports.__delete = __delete;
