// cncjs/dist/cncjs/server/api/api.mdi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.update = exports.read = exports.fetch = exports.create = exports.bulkUpdate = exports.__delete = void 0;

var _ensureArray = _interopRequireDefault(require("ensure-array"));

var _find = _interopRequireDefault(require("lodash/find"));

var _isPlainObject = _interopRequireDefault(require("lodash/isPlainObject"));

var _uuid = _interopRequireDefault(require("uuid"));

var _settings = _interopRequireDefault(require("../config/settings"));

var _logger = _interopRequireDefault(require("../lib/logger"));

var _configstore = _interopRequireDefault(require("../services/configstore"));

var _paging = require("./paging");

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var log = (0, _logger["default"])('api:mdi');
var CONFIG_KEY = 'mdi';

var getSanitizedRecords = function getSanitizedRecords() {
  var records = (0, _ensureArray["default"])(_configstore["default"].get(CONFIG_KEY, []));
  var shouldUpdate = false;

  for (var i = 0; i < records.length; ++i) {
    if (!(0, _isPlainObject["default"])(records[i])) {
      records[i] = {};
    }

    var record = records[i];

    if (!record.id) {
      record.id = _uuid["default"].v4();
      shouldUpdate = true;
    }
  }

  if (shouldUpdate) {
    log.debug("update sanitized records: ".concat(JSON.stringify(records))); // Pass `{ silent changes }` will suppress the change event

    _configstore["default"].set(CONFIG_KEY, records, {
      silent: true
    });
  }

  return records;
};

var fetch = function fetch(req, res) {
  var records = getSanitizedRecords();
  var paging = !!req.query.paging;

  if (paging) {
    var _req$query = req.query,
        _req$query$page = _req$query.page,
        page = _req$query$page === void 0 ? 1 : _req$query$page,
        _req$query$pageLength = _req$query.pageLength,
        pageLength = _req$query$pageLength === void 0 ? 10 : _req$query$pageLength;
    var totalRecords = records.length;

    var _getPagingRange = (0, _paging.getPagingRange)({
      page: page,
      pageLength: pageLength,
      totalRecords: totalRecords
    }),
        _getPagingRange2 = _slicedToArray(_getPagingRange, 2),
        begin = _getPagingRange2[0],
        end = _getPagingRange2[1];

    var pagedRecords = records.slice(begin, end);
    res.send({
      pagination: {
        page: Number(page),
        pageLength: Number(pageLength),
        totalRecords: Number(totalRecords)
      },
      records: pagedRecords.map(function (record) {
        var _record = _objectSpread({}, record),
            id = _record.id,
            name = _record.name,
            command = _record.command,
            _record$grid = _record.grid,
            grid = _record$grid === void 0 ? {} : _record$grid;

        return {
          id: id,
          name: name,
          command: command,
          grid: grid
        };
      })
    });
  } else {
    res.send({
      records: records.map(function (record) {
        var _record2 = _objectSpread({}, record),
            id = _record2.id,
            name = _record2.name,
            command = _record2.command,
            _record2$grid = _record2.grid,
            grid = _record2$grid === void 0 ? {} : _record2$grid;

        return {
          id: id,
          name: name,
          command: command,
          grid: grid
        };
      })
    });
  }
};

exports.fetch = fetch;

var create = function create(req, res) {
  var _req$body = _objectSpread({}, req.body),
      name = _req$body.name,
      command = _req$body.command,
      _req$body$grid = _req$body.grid,
      grid = _req$body$grid === void 0 ? {} : _req$body$grid;

  if (!name) {
    res.status(_constants.ERR_BAD_REQUEST).send({
      msg: 'The "name" parameter must not be empty'
    });
    return;
  }

  if (!command) {
    res.status(_constants.ERR_BAD_REQUEST).send({
      msg: 'The "command" parameter must not be empty'
    });
    return;
  }

  try {
    var records = getSanitizedRecords();
    var record = {
      id: _uuid["default"].v4(),
      name: name,
      command: command,
      grid: grid
    };
    records.push(record);

    _configstore["default"].set(CONFIG_KEY, records);

    res.send({
      err: null
    });
  } catch (err) {
    res.status(_constants.ERR_INTERNAL_SERVER_ERROR).send({
      msg: 'Failed to save ' + JSON.stringify(_settings["default"].cncrc)
    });
  }
};

exports.create = create;

var read = function read(req, res) {
  var id = req.params.id;
  var records = getSanitizedRecords();
  var record = (0, _find["default"])(records, {
    id: id
  });

  if (!record) {
    res.status(_constants.ERR_NOT_FOUND).send({
      msg: 'Not found'
    });
    return;
  }

  var _record3 = _objectSpread({}, record),
      name = _record3.name,
      command = _record3.command,
      _record3$grid = _record3.grid,
      grid = _record3$grid === void 0 ? {} : _record3$grid;

  res.send({
    id: id,
    name: name,
    command: command,
    grid: grid
  });
};

exports.read = read;

var update = function update(req, res) {
  var id = req.params.id;
  var records = getSanitizedRecords();
  var record = (0, _find["default"])(records, {
    id: id
  });

  if (!record) {
    res.status(_constants.ERR_NOT_FOUND).send({
      msg: 'Not found'
    });
    return;
  }

  var _req$body2 = _objectSpread({}, req.body),
      _req$body2$name = _req$body2.name,
      name = _req$body2$name === void 0 ? record.name : _req$body2$name,
      _req$body2$command = _req$body2.command,
      command = _req$body2$command === void 0 ? record.command : _req$body2$command,
      _req$body2$grid = _req$body2.grid,
      grid = _req$body2$grid === void 0 ? record.grid : _req$body2$grid;
  /*
  if (!name) {
      res.status(ERR_BAD_REQUEST).send({
          msg: 'The "name" parameter must not be empty'
      });
      return;
  }
   if (!command) {
      res.status(ERR_BAD_REQUEST).send({
          msg: 'The "command" parameter must not be empty'
      });
      return;
  }
  */


  try {
    record.name = String(name || '');
    record.command = String(command || '');
    record.grid = (0, _isPlainObject["default"])(grid) ? grid : {};

    _configstore["default"].set(CONFIG_KEY, records);

    res.send({
      err: null
    });
  } catch (err) {
    res.status(_constants.ERR_INTERNAL_SERVER_ERROR).send({
      msg: 'Failed to save ' + JSON.stringify(_settings["default"].cncrc)
    });
  }
};

exports.update = update;

var bulkUpdate = function bulkUpdate(req, res) {
  var _req$body3 = _objectSpread({}, req.body),
      records = _req$body3.records;

  if (!records) {
    res.status(_constants.ERR_BAD_REQUEST).send({
      msg: 'The "records" parameter must not be empty'
    });
    return;
  }

  var filteredRecords = (0, _ensureArray["default"])(records).filter(function (record) {
    return (0, _isPlainObject["default"])(record);
  });

  for (var i = 0; i < filteredRecords.length; ++i) {
    var record = filteredRecords[i];

    var _record4 = _objectSpread({}, record),
        id = _record4.id,
        name = _record4.name,
        command = _record4.command,
        _record4$grid = _record4.grid,
        grid = _record4$grid === void 0 ? {} : _record4$grid;

    if (!id) {
      record.id = _uuid["default"].v4();
    }

    record.name = String(name || '');
    record.command = String(command || '');
    record.grid = (0, _isPlainObject["default"])(grid) ? grid : {};
  }

  try {
    _configstore["default"].set(CONFIG_KEY, filteredRecords);

    res.send({
      err: null
    });
  } catch (err) {
    res.status(_constants.ERR_INTERNAL_SERVER_ERROR).send({
      msg: 'Failed to save ' + JSON.stringify(_settings["default"].cncrc)
    });
  }
};

exports.bulkUpdate = bulkUpdate;

var __delete = function __delete(req, res) {
  var id = req.params.id;
  var records = getSanitizedRecords();
  var record = (0, _find["default"])(records, {
    id: id
  });

  if (!record) {
    res.status(_constants.ERR_NOT_FOUND).send({
      msg: 'Not found'
    });
    return;
  }

  try {
    var filteredRecords = records.filter(function (record) {
      return record.id !== id;
    });

    _configstore["default"].set(CONFIG_KEY, filteredRecords);

    res.send({
      err: null
    });
  } catch (err) {
    res.status(_constants.ERR_INTERNAL_SERVER_ERROR).send({
      msg: 'Failed to save ' + JSON.stringify(_settings["default"].cncrc)
    });
  }
};

exports.__delete = __delete;
