// cncjs/dist/cncjs/server/api/api.watch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.readFile = exports.getFiles = void 0;

var _monitor = _interopRequireDefault(require("../services/monitor"));

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var getFiles = function getFiles(req, res) {
  var path = req.body.path || req.query.path || '';

  var files = _monitor["default"].getFiles(path);

  res.send({
    path: path,
    files: files
  });
};

exports.getFiles = getFiles;

var readFile = function readFile(req, res) {
  var file = req.body.file || req.query.file || '';

  _monitor["default"].readFile(file, function (err, data) {
    if (err) {
      if (err.code === 'ENOENT') {
        res.status(_constants.ERR_NOT_FOUND).send({
          msg: 'File not found'
        });
      } else {
        res.status(_constants.ERR_INTERNAL_SERVER_ERROR).send({
          msg: 'Failed reading file'
        });
      }

      return;
    }

    res.send({
      file: file,
      data: data
    });
  });
};

exports.readFile = readFile;
