// cncjs/dist/cncjs/server/controllers/TinyG/TinyGLineParserResultQueueReports.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _get2 = _interopRequireDefault(require("lodash/get"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var TinyGLineParserResultQueueReports = /*#__PURE__*/function () {
  function TinyGLineParserResultQueueReports() {
    _classCallCheck(this, TinyGLineParserResultQueueReports);
  }

  _createClass(TinyGLineParserResultQueueReports, null, [{
    key: "parse",
    value: function parse(data) {
      var qr = (0, _get2["default"])(data, 'r.qr') || (0, _get2["default"])(data, 'qr');
      var qi = (0, _get2["default"])(data, 'r.qi') || (0, _get2["default"])(data, 'qi');
      var qo = (0, _get2["default"])(data, 'r.qo') || (0, _get2["default"])(data, 'qo');

      if (!qr) {
        return null;
      }

      var payload = {
        qr: Number(qr) || 0,
        qi: Number(qi) || 0,
        qo: Number(qo) || 0
      };
      return {
        type: TinyGLineParserResultQueueReports,
        payload: payload
      };
    }
  }]);

  return TinyGLineParserResultQueueReports;
}();

var _default = TinyGLineParserResultQueueReports;
exports["default"] = _default;
