// cncjs/dist/cncjs/server/constants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ERR_UNAUTHORIZED = exports.ERR_PRECONDITION_FAILED = exports.ERR_PAYLOAD_TOO_LARGE = exports.ERR_NOT_FOUND = exports.ERR_NOT_ACCEPTABLE = exports.ERR_METHOD_NOT_ALLOWED = exports.ERR_LENGTH_REQUIRED = exports.ERR_INTERNAL_SERVER_ERROR = exports.ERR_FORBIDDEN = exports.ERR_CONFLICT = exports.ERR_BAD_REQUEST = void 0;
// Error Codes
var ERR_BAD_REQUEST = 400;
exports.ERR_BAD_REQUEST = ERR_BAD_REQUEST;
var ERR_UNAUTHORIZED = 401;
exports.ERR_UNAUTHORIZED = ERR_UNAUTHORIZED;
var ERR_FORBIDDEN = 403;
exports.ERR_FORBIDDEN = ERR_FORBIDDEN;
var ERR_NOT_FOUND = 404;
exports.ERR_NOT_FOUND = ERR_NOT_FOUND;
var ERR_METHOD_NOT_ALLOWED = 405;
exports.ERR_METHOD_NOT_ALLOWED = ERR_METHOD_NOT_ALLOWED;
var ERR_NOT_ACCEPTABLE = 406;
exports.ERR_NOT_ACCEPTABLE = ERR_NOT_ACCEPTABLE;
var ERR_CONFLICT = 409;
exports.ERR_CONFLICT = ERR_CONFLICT;
var ERR_LENGTH_REQUIRED = 411;
exports.ERR_LENGTH_REQUIRED = ERR_LENGTH_REQUIRED;
var ERR_PRECONDITION_FAILED = 412;
exports.ERR_PRECONDITION_FAILED = ERR_PRECONDITION_FAILED;
var ERR_PAYLOAD_TOO_LARGE = 413;
exports.ERR_PAYLOAD_TOO_LARGE = ERR_PAYLOAD_TOO_LARGE;
var ERR_INTERNAL_SERVER_ERROR = 500;
exports.ERR_INTERNAL_SERVER_ERROR = ERR_INTERNAL_SERVER_ERROR;
