// @trendmicro/react-iframe/lib/Iframe.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /* eslint jsx-a11y/iframe-has-title: 0 */


var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var mapSandboxToString = function mapSandboxToString() {
    var sandbox = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

    if (typeof sandbox === 'string') {
        return sandbox;
    }
    sandbox = _extends({}, sandbox);
    return Object.keys(sandbox).filter(function (key) {
        return sandbox[key];
    }).map(function (s) {
        return s.replace(/[A-Z]/g, '-$&').toLowerCase();
    }).join(' ');
};

var Iframe = (_temp = _class = function (_PureComponent) {
    _inherits(Iframe, _PureComponent);

    function Iframe() {
        _classCallCheck(this, Iframe);

        return _possibleConstructorReturn(this, (Iframe.__proto__ || Object.getPrototypeOf(Iframe)).apply(this, arguments));
    }

    _createClass(Iframe, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                style = _props.style,
                props = _objectWithoutProperties(_props, ['style']);

            if (props.sandbox === false) {
                delete props.sandbox;
            } else if (_typeof(props.sandbox) === 'object') {
                props.sandbox = mapSandboxToString(props.sandbox);
            }

            return _react2.default.createElement('iframe', _extends({}, props, {
                style: _extends({
                    borderWidth: 0
                }, style)
            }));
        }
    }]);

    return Iframe;
}(_react.PureComponent), _class.propTypes = {
    sandbox: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.string, _propTypes2.default.shape({
        // Re-enables form submission
        allowForms: _propTypes2.default.bool,

        // Sandboxed frames will block modal dialogs by default
        allowModals: _propTypes2.default.bool,

        // Re-enables APIs
        allowPointerLock: _propTypes2.default.bool,

        // Re-enables popups
        allowPopups: _propTypes2.default.bool,

        // Allows the iframe content to be treated as being from the same origin
        allowSameOrigin: _propTypes2.default.bool,

        // Re-enables scripts
        allowScripts: _propTypes2.default.bool,

        // Allows the iframe content to navigate its top-level browsing context
        allowTopNavigation: _propTypes2.default.bool
    })])
}, _class.defaultProps = {
    width: '100%',
    height: '100%',
    sandbox: {
        allowForms: true,
        allowModals: true,
        allowPointerLock: false,
        allowPopups: true,
        allowSameOrigin: true,
        allowScripts: true,
        allowTopNavigation: false
    }
}, _temp);
exports.default = Iframe;
