// cncjs/dist/cncjs/server/config/settings.production.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _crypto = _interopRequireDefault(require("crypto"));

var _os = _interopRequireDefault(require("os"));

var _path = _interopRequireDefault(require("path"));

var _package = _interopRequireDefault(require("../../package.json"));

var _urljoin = _interopRequireDefault(require("../lib/urljoin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var publicPath = function (payload) {
  var algorithm = 'sha1';
  var buf = String(payload);

  var hash = _crypto["default"].createHash(algorithm).update(buf).digest('hex');

  return '/' + hash.substr(0, 8) + '/'; // 8 digits
}(_package["default"].version);

var maxAge = 365 * 24 * 60 * 60 * 1000; // one year

var _default = {
  route: '/',
  // with trailing slash
  assets: {
    app: {
      routes: [// with trailing slash
      (0, _urljoin["default"])(publicPath, '/'), '/' // fallback
      ],
      path: _path["default"].resolve(__dirname, '..', '..', 'app'),
      maxAge: maxAge
    }
  },
  backend: {
    enable: false,
    // disable backend service in production
    host: 'localhost',
    port: 80,
    route: 'api/'
  },
  cluster: {
    // note. node-inspector cannot debug child (forked) process
    enable: false,
    maxWorkers: _os["default"].cpus().length || 1
  },
  winston: {
    // https://github.com/winstonjs/winston#logging-levels
    level: 'info'
  }
};
exports["default"] = _default;
